pkg/virt-handler/vm.go

```diff
@@ -308,27 +310,54 @@ func (d *VirtualMachineController) execute(key string) error {
 		return err
 	}
 
+	log.Log.V(3).Infof("Processing vmi %v, existing: %v\n", vmi.Name, vmiExists)
+	if vmiExists {
+		log.Log.V(3).Infof("vmi is in phase: %v\n", vmi.Status.Phase)
+	}
+
 	domain, domainExists, err := d.getDomainFromCache(key)
 	if err != nil {
 		return err
 	}
 
+	// Ignore domains from an older VMI
+	if vmiExists && domainExists && domain.Spec.Metadata.KubeVirt.UID != vmi.UID {
+		log.Log.Object(vmi).Info("Ignoring domain from an older VMI, will be handled by its own VMI.")
+		domain = nil
+		domainExists = false
```

Comment:
In this case you can simply return.

Reply:
I don't think so: when you stop and start VMs very fast, there is a high chance that you get the new VMI but an old domain here. If you just return, no new domain would be created for the new VMI.

Reply:
I think we'll need to return early if this occurs.

This shouldn't happen under normal operation. It would be nice to understand what race condition is actually causing this, and close the gap there.  

The sequence should be.
1. vmi is marked for deletion
2. domain is signaled for shutdown.
3. domain is shutdown or destroyed
4. vmi's status is updated to finalized state.
5a. virt-controller removes VMI finalizer once Pod is finalized
5.b virt-handler processes deleted domain.

I'm guessing the race condition is 5a completing before 5b, which leaves a small window for another VMI to get created before the domain is removed from virt-handler's cache.

Reply:
Yes I think so too. But we could return with an error (so non-nil) so that the controller retries. The window should be really small, since the old vmi can only disappear from the cluster if its pod also disappeared. So the domain should already be gone at the repost-stage or it should be gone for as soon as ourch file-based watchdog cleans the cache up.

Reply:
We should definitely log an error here, but we shouldn't need to return an err to get the retry. The controller should retry automatically when the domain is deleted. The same VMI key will be queued when the watchdog expires for the old domain.

Reply:
returning an error sounds like an easy solution, will try that. thx!

Reply:
@davidvossel is right. Just log the error and return it with `nil`. Once the domain gets deleted the controller will be triggered again.